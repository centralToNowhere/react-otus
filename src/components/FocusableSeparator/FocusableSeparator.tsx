import React, { FC, useEffect, useState } from "react";
import styled from "@emotion/styled";
import { BREAKPOINTS, COLORS } from "@/styles/ui-styled";

const resizerSize = 30;

export const FocusableSeparator: FC = () => {
  const [valueNow, setValueNow] = useState(50);
  const refContainerEl = React.useRef<HTMLDivElement | null>(null);
  const pointerDx = React.useRef<number | null>(null);

  useEffect(() => {
    let parentY: number | null = null,
      parentH: number | null = null;

    const onPointerStart = (e: MouseEvent) => {
      const pointerY = e.clientY;

      if (!refContainerEl.current) {
        return;
      }

      const resizerContainer = refContainerEl.current;
      const resizerRect = resizerContainer.getBoundingClientRect();

      pointerDx.current = pointerY - resizerRect.y;

      const parent = resizerContainer.parentElement as HTMLDivElement;

      const { y, height } = parent.getBoundingClientRect();
      [parentY, parentH] = [y, height];

      parent.addEventListener("mousemove", onPointerMove);
    };

    const onPointerMove = (e: MouseEvent) => {
      e.preventDefault();
      const pointerY = e.clientY;

      if (
        parentY === null ||
        parentH === null ||
        pointerY === null ||
        !refContainerEl.current
      ) {
        return;
      }

      const resizerContainer = refContainerEl.current;
      const previousElement =
        resizerContainer.previousElementSibling as HTMLElement;
      const nextElement = resizerContainer.nextElementSibling as HTMLElement;

      if (!previousElement || !nextElement) {
        return;
      }

      const dx = pointerDx.current ? pointerDx.current : 0;
      let resizerTop: number;

      if (pointerY - dx <= parentY) {
        resizerTop = 0;
      } else if (pointerY + resizerSize - dx >= parentY + parentH) {
        resizerTop = parentY + parentH - resizerSize;
      } else {
        resizerTop = pointerY - dx;
      }

      const heightPrev = resizerTop - parentY;
      const heightNext = parentH - resizerTop - resizerSize;
      const valueNow = Math.round((resizerTop / (parentH - resizerSize)) * 100);

      setValueNow(valueNow);

      previousElement.setAttribute("data-testh", String(valueNow));
      previousElement.style.height = heightPrev + "px";
      nextElement.setAttribute("data-testh", String(100 - valueNow));
      nextElement.style.height = heightNext + "px";
    };

    const onPointerEnd = () => {
      if (!refContainerEl.current) {
        return;
      }

      const resizerContainer = refContainerEl.current;
      const parent = resizerContainer.parentElement as HTMLDivElement;

      parentY = null;
      parentH = null;

      parent.removeEventListener("mousemove", onPointerMove);
    };

    if (!refContainerEl.current) {
      return;
    }

    const resizerContainer = refContainerEl.current;

    resizerContainer.addEventListener("mousedown", onPointerStart);
    document.addEventListener("mouseup", onPointerEnd);

    return () => {
      resizerContainer.removeEventListener("mousedown", onPointerStart);
      resizerContainer.removeEventListener("mousemove", onPointerMove);
      document.removeEventListener("mouseup", onPointerEnd);
    };
  }, []);

  return (
    <StyledResizerLine
      ref={refContainerEl}
      role="separator"
      aria-valuenow={valueNow}
    >
      <div className={"handle-icon"} />
    </StyledResizerLine>
  );
};

const StyledResizerLine = styled.div`
  width: 100%;
  height: ${resizerSize}px;
  cursor: grabbing;
  user-select: none;
  overflow: hidden;
  border-top: 2px solid ${COLORS.border};

  &:hover {
    opacity: 0.7;
  }

  .handle-icon {
    display: block;
    background: ${COLORS.border};
    width: 50px;
    margin: 10px auto 0 auto;
    height: 6px;
    border-radius: 10px;
  }

  @media screen and (max-width: ${BREAKPOINTS.xl}) {
    display: none;
  }
`;
