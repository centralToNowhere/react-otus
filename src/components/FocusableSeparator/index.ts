export { FocusableSeparator } from "@/components/FocusableSeparator/FocusableSeparator";
