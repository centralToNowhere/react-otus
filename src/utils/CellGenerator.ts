import { ICell } from "@/Cell/Cell";

const willBeAlive = (isCellStateAlive: 1 | 0, neighbours: number): boolean => {
  if (isCellStateAlive) {
    if (neighbours === 2 || neighbours === 3) {
      return true;
    }
  } else {
    if (neighbours === 3) {
      return true;
    }
  }

  return false;
};

export const getNextGeneration = (
  cellsInRow: number,
  cellsInCol: number,
  indexedCells: Array<1 | 0>
): ICell[] => {
  return indexedCells.reduce((indexed: ICell[], cellState, i) => {
    if (
      willBeAlive(
        cellState,
        calculateNeighbours(cellState, indexedCells, i, cellsInRow)
      )
    ) {
      indexed.push({
        x: i % cellsInRow,
        y: Math.floor(i / cellsInRow),
      });
    }

    return indexed;
  }, []);
};

const calculateNeighbours = (
  cellState: 1 | 0,
  indexedCells: Array<1 | 0>,
  i: number,
  cellsInRow: number
) => {
  let neighbours = 0;

  // Некрасиво, но производительно

  // если ячейка находится на границе поля
  // right
  if ((i + 1) % cellsInRow !== 0) {
    neighbours += indexedCells[i + 1];
  }

  // left
  if (i % cellsInRow !== 0) {
    neighbours += indexedCells[i - 1];
  }

  // top
  if (i - cellsInRow >= 0) {
    neighbours += indexedCells[i - cellsInRow];

    // top right
    if ((i + 1) % cellsInRow !== 0) {
      neighbours += indexedCells[i + 1 - cellsInRow];
    }

    // top left
    if (i % cellsInRow !== 0) {
      neighbours += indexedCells[i - 1 - cellsInRow];
    }
  }

  // bottom
  if (i + cellsInRow < indexedCells.length) {
    neighbours += indexedCells[i + cellsInRow];

    // bottom right
    if ((i + 1) % cellsInRow !== 0) {
      neighbours += indexedCells[i + 1 + cellsInRow];
    }

    // bottom left
    if (i % cellsInRow !== 0) {
      neighbours += indexedCells[i - 1 + cellsInRow];
    }
  }
  // if (
  //   (i + 1) % cellsInRow === 0 ||
  //   i % cellsInRow === 0 ||
  //   i - cellsInRow < 0 ||
  //   i + cellsInRow >= indexedCells.length
  // ) {
  //
  //   }
  // } else {
  //   neighbours += indexedCells[i + 1];
  //   neighbours += indexedCells[i - 1];
  //   neighbours += indexedCells[i - cellsInRow];
  //   neighbours += indexedCells[i + 1 - cellsInRow];
  //   neighbours += indexedCells[i - 1 - cellsInRow];
  //   neighbours += indexedCells[i + cellsInRow];
  //   neighbours += indexedCells[i + 1 + cellsInRow];
  //   neighbours += indexedCells[i - 1 + cellsInRow];
  // }

  return neighbours;
};

export const getRandomCells = (
  cellsInRow: number,
  cellsInCol: number,
  percentage: number
): ICell[] => {
  return getAllCells(cellsInCol, cellsInRow, (x: number, y: number) => {
    if (Math.random() < percentage) {
      return {
        x,
        y,
      };
    }

    return null;
  });
};

export const getInitialCells = (
  rowWidth: number,
  colHeight: number,
  cellSize: number
): ICell[] => {
  return getAllCells(
    Math.floor(colHeight / cellSize),
    Math.floor(rowWidth / cellSize),
    (x: number, y: number) => {
      if ((x - y) % 4 === 0 || (x + y) % 4 === 0) {
        return {
          x,
          y,
        };
      }

      return null;
    }
  );
};

const getAllCells = (
  cellsInCol: number,
  cellsInRow: number,
  getCellOrNull: (x: number, y: number) => ICell | null
) => {
  return Array.from(Array(cellsInCol)).reduce(
    (allCells, cell, rowNumber): ICell[] => {
      allCells = allCells.concat(
        getRowCells(cellsInRow, rowNumber, getCellOrNull)
      );

      return allCells;
    },
    []
  );
};

const getRowCells = (
  cellsInRow: number,
  rowNumber: number,
  getCellOrNull: (x: number, y: number) => ICell | null
) => {
  return Array.from(Array(cellsInRow)).reduce(
    (rowCells, cell, colNumber): ICell[] => {
      const newCell = getCellOrNull(colNumber, rowNumber);

      if (newCell !== null) {
        rowCells.push(newCell);
      }

      return rowCells;
    },
    []
  );
};
