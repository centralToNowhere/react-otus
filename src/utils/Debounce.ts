import { useMemo } from "react";

export interface Debounced<T> {
  (value: T): void;
  clear: () => void;
}

export const useDebounce = <T>(
  callback: (value: T) => void,
  delay = 1000
): Debounced<T> => {
  // useMemo for preventing timeout clearing on every character input
  return useMemo<Debounced<T>>(() => {
    let timeout: ReturnType<typeof window.setTimeout> | null = null;

    const clearDebounce = () => {
      if (timeout !== null) {
        clearTimeout(timeout);
        timeout = null;
      }
    };

    const debounced: Debounced<T> = (value: T) => {
      clearDebounce();

      timeout = setTimeout(() => {
        clearDebounce();
        callback(value);
      }, delay);
    };

    debounced.clear = clearDebounce;

    return debounced;
  }, [callback, delay]);
};
