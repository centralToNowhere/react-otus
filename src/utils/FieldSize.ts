export const fieldWidth = window.innerWidth;
export const fieldHeight = window.innerHeight / 2;
