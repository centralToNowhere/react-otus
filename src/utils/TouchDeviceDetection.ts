export function isTouchDevice(): boolean {
  return (
    "ontouchstart" in window ||
    window.navigator.maxTouchPoints > 0 ||
    (
      window.navigator as Navigator & {
        msMaxTouchPoints: number;
      }
    ).msMaxTouchPoints > 0
  );
}
