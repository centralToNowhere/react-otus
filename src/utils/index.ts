export {
  isValidNumericString,
  isValidCellSizeString,
  isValidPositiveNumericString,
  isValidNonNegativeNumericString,
} from "./Validators";
export { useDebounce } from "./Debounce";
export { fieldWidth, fieldHeight } from "./FieldSize";
export type { Debounced } from "./Debounce";
export type { DeepPartial } from "./ts-types";
