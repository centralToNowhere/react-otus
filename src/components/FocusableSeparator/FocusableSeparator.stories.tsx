import React from "react";
import { Story, Meta } from "@storybook/react";
import { FocusableSeparator } from "@/components/FocusableSeparator";
import { css } from "@emotion/react";

export default {
  title: "Components/FocusableSeparator",
  component: FocusableSeparator,
} as Meta;

const Template: Story = () => {
  return (
    <div
      css={css`
        width: 100px;
        position: absolute;
        top: 100px;
        left: 100px;
        overflow: hidden;
      `}
    >
      <div
        css={css`
          background: green;
          overflow: hidden;
        `}
      >
        Content of div 1
      </div>
      <FocusableSeparator />
      <div
        css={css`
          background: yellow;
          overflow: hidden;
        `}
      >
        Content of div 2
      </div>
    </div>
  );
};

export const FocusableSeparatorDefault = Template;
