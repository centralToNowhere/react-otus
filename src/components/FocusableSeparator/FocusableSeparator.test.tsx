import React, { FC, MutableRefObject } from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { FocusableSeparator } from "@/components/FocusableSeparator/FocusableSeparator";
import { css } from "@emotion/react";
import { isTouchDevice } from "@/utils/TouchDeviceDetection";

const parentBox = {
  bottom: 130,
  height: 130,
  left: 0,
  right: 100,
  top: 0,
  width: 100,
  x: 0,
  y: 0,
};

const parentBoxWithOffset = {
  bottom: 140,
  height: 130,
  left: 0,
  right: 100,
  top: 10,
  width: 100,
  x: 0,
  y: 10,
};

const handleBox = {
  bottom: 80,
  height: 30,
  left: 0,
  right: 100,
  top: 50,
  width: 100,
  x: 0,
  y: 50,
};

const handleBoxWithDx = {
  bottom: 65,
  height: 30,
  left: 0,
  right: 100,
  top: 35,
  width: 100,
  x: 0,
  y: 35,
};

const Template: FC = () => {
  return (
    <div
      css={css`
        width: 100px;
        overflow: hidden;
      `}
      data-testid="div-parent"
    >
      <div
        css={css`
          background: green;
          overflow: hidden;
        `}
        data-testid="div1"
        data-testh="50"
      >
        Content of div 1
      </div>
      <FocusableSeparator />
      <div
        css={css`
          background: yellow;
          overflow: hidden;
        `}
        data-testid="div2"
        data-testh="50"
      >
        Content of div 2
      </div>
    </div>
  );
};

const NoSiblings: FC = () => {
  return (
    <div
      css={css`
        width: 100px;
        overflow: hidden;
      `}
      data-testid="div-parent"
    >
      <FocusableSeparator />
    </div>
  );
};

const OneSiblingPrev: FC = () => {
  return (
    <div
      css={css`
        width: 100px;
        overflow: hidden;
      `}
      data-testid="div-parent"
    >
      <div
        css={css`
          background: green;
          overflow: hidden;
        `}
        data-testid="div1"
        data-testh="50"
      >
        Content of div 1
      </div>
      <FocusableSeparator />
    </div>
  );
};

const OneSiblingNext: FC = () => {
  return (
    <div
      css={css`
        width: 100px;
        overflow: hidden;
      `}
      data-testid="div-parent"
    >
      <FocusableSeparator />
      <div
        css={css`
          background: yellow;
          overflow: hidden;
        `}
        data-testid="div2"
        data-testh="50"
      >
        Content of div 2
      </div>
    </div>
  );
};

beforeEach(() => {
  jest.restoreAllMocks();
});

describe("focusable separator tests", () => {
  it("should render focusable separator", () => {
    const { asFragment } = render(<FocusableSeparator />);

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole("separator")).toBeInTheDocument();
  });

  it("from 50/50 to 70/30", async () => {
    render(<Template />);

    const handle = screen.getByRole("separator");

    const parent = screen.getByTestId("div-parent");
    const div1 = screen.getByTestId("div1");
    const div2 = screen.getByTestId("div2");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );
    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 70,
      })
    );

    fireEvent.mouseUp(document);

    expect(handle).toHaveAttribute("aria-valuenow", "70");
    expect(div1).toHaveAttribute("data-testh", "70");
    expect(div2).toHaveAttribute("data-testh", "30");
    expect(div1).toHaveStyle({
      height: "70px",
    });
    expect(div2).toHaveStyle({
      height: "30px",
    });
  });

  it("should resize from 50/50 to 100/0", () => {
    render(<Template />);

    const handle = screen.getByRole("separator");

    const parent = screen.getByTestId("div-parent");
    const div1 = screen.getByTestId("div1");
    const div2 = screen.getByTestId("div2");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );
    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 100,
      })
    );

    fireEvent.mouseUp(document);

    expect(handle).toHaveAttribute("aria-valuenow", "100");
    expect(div1).toHaveAttribute("data-testh", "100");
    expect(div2).toHaveAttribute("data-testh", "0");
    expect(div1).toHaveStyle({
      height: "100px",
    });
    expect(div2).toHaveStyle({
      height: "0px",
    });
  });

  it("should resize from 50/50 to 0/100", () => {
    render(<Template />);

    const handle = screen.getByRole("separator");

    const parent = screen.getByTestId("div-parent");
    const div1 = screen.getByTestId("div1");
    const div2 = screen.getByTestId("div2");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );
    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 0,
      })
    );

    fireEvent.mouseUp(document);

    expect(handle).toHaveAttribute("aria-valuenow", "0");
    expect(div1).toHaveAttribute("data-testh", "0");
    expect(div2).toHaveAttribute("data-testh", "100");
    expect(div1).toHaveStyle({
      height: "0px",
    });
    expect(div2).toHaveStyle({
      height: "100px",
    });
  });

  describe("with offset", () => {
    it("30/70", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 60,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 65,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "30");
      expect(div1).toHaveAttribute("data-testh", "30");
      expect(div2).toHaveAttribute("data-testh", "70");
      expect(div1).toHaveStyle({
        height: "30px",
      });
      expect(div2).toHaveStyle({
        height: "70px",
      });
    });

    it("70/30", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 60,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 105,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "70");
      expect(div1).toHaveAttribute("data-testh", "70");
      expect(div2).toHaveAttribute("data-testh", "30");
      expect(div1).toHaveStyle({
        height: "70px",
      });
      expect(div2).toHaveStyle({
        height: "30px",
      });
    });

    it("pointerY < parentY", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: -50,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "0");
      expect(div1).toHaveAttribute("data-testh", "0");
      expect(div2).toHaveAttribute("data-testh", "100");
    });

    it("pointerY - dx === parentY", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 25,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "0");
      expect(div1).toHaveAttribute("data-testh", "0");
      expect(div2).toHaveAttribute("data-testh", "100");
    });

    it("pointerY - dx < parentY", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 24,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "0");
      expect(div1).toHaveAttribute("data-testh", "0");
      expect(div2).toHaveAttribute("data-testh", "100");
    });

    it("pointerY - dx > parentY", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 26,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "1");
      expect(div1).toHaveAttribute("data-testh", "1");
      expect(div2).toHaveAttribute("data-testh", "99");
    });

    it("pointerY + resizerSize - dx > parentY + parent height", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 126,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "100");
      expect(div1).toHaveAttribute("data-testh", "100");
      expect(div2).toHaveAttribute("data-testh", "0");
    });

    it("pointerY + resizerSize - dx === parentY + parent height", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 125,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "100");
      expect(div1).toHaveAttribute("data-testh", "100");
      expect(div2).toHaveAttribute("data-testh", "0");
    });

    it("pointerY + resizerSize - dx < parentY + parent height", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest
        .fn()
        .mockReturnValue(parentBoxWithOffset);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBoxWithDx);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 124,
        })
      );

      fireEvent.mouseUp(document);

      expect(handle).toHaveAttribute("aria-valuenow", "99");
      expect(div1).toHaveAttribute("data-testh", "99");
      expect(div2).toHaveAttribute("data-testh", "1");
    });
  });

  it("should not resize - no elements inside parent container", () => {
    render(<NoSiblings />);

    const handle = screen.getByRole("separator");
    const parent = screen.getByTestId("div-parent");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );

    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 25,
      })
    );

    expect(handle).toHaveAttribute("aria-valuenow", "50");
  });

  it("should not resize - one element inside container (before separator)", () => {
    render(<OneSiblingPrev />);

    const handle = screen.getByRole("separator");
    const parent = screen.getByTestId("div-parent");
    const div1 = screen.getByTestId("div1");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );

    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 25,
      })
    );

    expect(handle).toHaveAttribute("aria-valuenow", "50");
    expect(div1).toHaveAttribute("data-testh", "50");
  });

  it("should not resize - one element inside container (after separator)", () => {
    render(<OneSiblingNext />);

    const handle = screen.getByRole("separator");
    const parent = screen.getByTestId("div-parent");
    const div2 = screen.getByTestId("div2");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );

    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 25,
      })
    );

    expect(handle).toHaveAttribute("aria-valuenow", "50");
    expect(div2).toHaveAttribute("data-testh", "50");
  });

  it("should not resize - resizerRef === null", () => {
    jest.spyOn(React, "useRef").mockReturnValue({
      current: null,
      set: () => null,
    } as MutableRefObject<HTMLDivElement | null>);

    render(<Template />);

    const handle = screen.getByRole("separator");
    const parent = screen.getByTestId("div-parent");
    const div1 = screen.getByTestId("div1");
    const div2 = screen.getByTestId("div2");

    parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
    handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

    fireEvent(
      handle,
      new MouseEvent("mousedown", {
        clientY: 50,
      })
    );
    fireEvent(
      parent,
      new MouseEvent("mousemove", {
        clientY: 25,
      })
    );

    fireEvent.mouseUp(document);

    expect(handle).toHaveAttribute("aria-valuenow", "50");
    expect(div1).toHaveAttribute("data-testh", "50");
    expect(div2).toHaveAttribute("data-testh", "50");
  });

  describe("on pointer end", () => {
    it("should not resize after mouseup", () => {
      render(<Template />);

      const handle = screen.getByRole("separator");

      const parent = screen.getByTestId("div-parent");
      const div1 = screen.getByTestId("div1");
      const div2 = screen.getByTestId("div2");

      parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );
      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 30,
        })
      );

      fireEvent.mouseUp(document);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );

      expect(handle).toHaveAttribute("aria-valuenow", "30");
      expect(div1).toHaveAttribute("data-testh", "30");
      expect(div2).toHaveAttribute("data-testh", "70");
    });
  });

  describe("on unmount", () => {
    it("should not call onPointerEnd on document mouseup after unmount", () => {
      const { unmount } = render(<Template />);

      const mockRemoveEventListener = jest.spyOn(
        document,
        "removeEventListener"
      );

      const handle = screen.getByRole("separator");
      const parent = screen.getByTestId("div-parent");

      parent.getBoundingClientRect = jest.fn().mockReturnValue(parentBox);
      handle.getBoundingClientRect = jest.fn().mockReturnValue(handleBox);

      fireEvent(
        handle,
        new MouseEvent("mousedown", {
          clientY: 50,
        })
      );

      fireEvent(
        parent,
        new MouseEvent("mousemove", {
          clientY: 25,
        })
      );

      unmount();

      expect(mockRemoveEventListener).toHaveBeenCalledWith(
        "mouseup",
        expect.any(Function)
      );
    });
  });

  describe("isTouchDevice", () => {
    beforeEach(() => {
      jest.restoreAllMocks();
    });

    it("should return true", () => {
      let mockedNavigator: Navigator & {
        msMaxTouchPoints: number;
      };

      const windowMock = jest.spyOn(window, "window", "get");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { ontouchstart, ...mockedWindow } = window;

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: -1,
        msMaxTouchPoints: 1,
      };

      windowMock.mockReturnValue({
        ...mockedWindow,
        navigator: mockedNavigator as Navigator,
      } as Window & typeof globalThis);

      expect(isTouchDevice()).toBe(true);

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: 1,
        msMaxTouchPoints: -1,
      };

      windowMock.mockReturnValue({
        ...mockedWindow,
        navigator: mockedNavigator as Navigator,
      } as Window & typeof globalThis);

      expect(isTouchDevice()).toBe(true);

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: 0,
        msMaxTouchPoints: 1,
      };

      windowMock.mockReturnValue({
        ...mockedWindow,
        navigator: mockedNavigator as Navigator,
      } as Window & typeof globalThis);

      expect(isTouchDevice()).toBe(true);

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: 1,
        msMaxTouchPoints: 0,
      };

      windowMock.mockReturnValue({
        ...mockedWindow,
        navigator: mockedNavigator as Navigator,
      } as Window & typeof globalThis);

      expect(isTouchDevice()).toBe(true);

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: -1,
        msMaxTouchPoints: -1,
      };

      windowMock.mockReturnValue({
        ...window,
        ontouchstart: () => {
          // empty
        },
        navigator: mockedNavigator as Navigator,
      } as Window &
        typeof globalThis & {
          ontouchstart: () => void;
        });

      expect(isTouchDevice()).toBe(true);

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: 0,
        msMaxTouchPoints: 0,
      };

      windowMock.mockReturnValue({
        ...window,
        ontouchstart: () => {
          // empty
        },
        navigator: mockedNavigator as Navigator,
      } as Window &
        typeof globalThis & {
          ontouchstart: () => void;
        });

      expect(isTouchDevice()).toBe(true);
    });

    it("should return false", () => {
      let mockedNavigator: Navigator & {
        msMaxTouchPoints: number;
      };

      const windowMock = jest.spyOn(window, "window", "get");
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { ontouchstart, ...mockedWindow } = window;

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: -1,
        msMaxTouchPoints: -1,
      };

      windowMock.mockReturnValue({
        ...mockedWindow,
        navigator: mockedNavigator as Navigator,
      } as Window & typeof globalThis);

      expect(isTouchDevice()).toBe(false);

      mockedNavigator = {
        ...window.navigator,
        maxTouchPoints: 0,
        msMaxTouchPoints: 0,
      };

      windowMock.mockReturnValue({
        ...mockedWindow,
        navigator: mockedNavigator as Navigator,
      } as Window & typeof globalThis);

      expect(isTouchDevice()).toBe(false);
    });
  });
});
