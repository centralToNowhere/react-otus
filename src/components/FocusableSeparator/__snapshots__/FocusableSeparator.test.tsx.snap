// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`focusable separator tests should render focusable separator 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
  height: 30px;
  cursor: -webkit-grabbing;
  cursor: grabbing;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  border-top: 2px solid #6a6d75;
}

.emotion-0:hover {
  opacity: 0.7;
}

.emotion-0 .handle-icon {
  display: block;
  background: #6a6d75;
  width: 50px;
  margin: 10px auto 0 auto;
  height: 6px;
  border-radius: 10px;
}

<div
    aria-valuenow="50"
    class="emotion-0 emotion-1"
    role="separator"
  >
    <div
      class="handle-icon"
    />
  </div>
</DocumentFragment>
`;
